library/trans.c

```diff
@@ -557,9 +557,88 @@ int dnet_trans_iterate_move_transaction(struct dnet_net_state *st, struct list_h
 	return trans_moved;
 }
 
+struct dnet_ping_node_private
+{
+	struct dnet_session	*session;
+	struct dnet_net_state	*net_state;
+};
+
+static int dnet_ping_stall_node_complete(struct dnet_addr *addr __unused, struct dnet_cmd *cmd, void *priv)
+{
+	struct dnet_ping_node_private *ping_private;
+	struct dnet_net_state *st;
+
+	if (!is_trans_destroyed(cmd)) {
+	        ping_private = priv;
+		st = ping_private->net_state;
+
+		dnet_session_destroy(ping_private->session);
```

Comment:
Callback can and should be invoked multiple times - consider operations which are being split on server, or multiple replies (some data and status, or multiple data chunks like per-backend status and ack showing transaction has been completed). Your check says 'call this chunk when transaction is not about to be destroyed', which means potentially calling it multiple times.

You have removed dnet_update_status() function and its callback, which implemented correct reference counter mechanism which works for multiple replies. Your new code should use similar logic otherwise it is prone to crash.